pkg/controller/deployment/util/deployment_util.go

```diff
@@ -691,6 +800,69 @@ func IsRollingUpdate(deployment *extensions.Deployment) bool {
 	return deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType
 }
 
+// IsDeploymentComplete considers a deployment as complete once its desired replicas equals its
+// updatedReplicas and it doesn't violate minimum availability.
+func IsDeploymentComplete(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	return newStatus.UpdatedReplicas == deployment.Spec.Replicas &&
+		newStatus.AvailableReplicas >= deployment.Spec.Replicas-MaxUnavailable(*deployment)
+}
+
+// IsDeploymentProgressing reports if a deployment is progressing.
+func IsDeploymentProgressing(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	return deployment.Status.Replicas != newStatus.Replicas ||
+		deployment.Status.UpdatedReplicas != newStatus.UpdatedReplicas ||
+		deployment.Status.AvailableReplicas != newStatus.AvailableReplicas ||
+		deployment.Status.UnavailableReplicas != newStatus.UnavailableReplicas
```

Comment:
I already have another helper for detecting lack of progress: `IsDeploymentFailed`. That being said, the way progress is estimated in `syncDeploymentStatus` needs a bit more thought.